stratosphere/sm/source/sm_service_manager.cpp

```diff
@@ -0,0 +1,723 @@
+            /* Send command to query if we should mitm. */
+            {
+                IpcCommand c;
+                ipcInitialize(&c);
+                struct {
+                    u64 magic;
+                    u64 cmd_id;
+                    u64 pid;
+                } *info = ((decltype(info))ipcPrepareHeader(&c, sizeof(*info)));
+                info->magic = SFCI_MAGIC;
+                info->cmd_id = 65000;
+                info->pid = pid;
+                R_TRY(ipcDispatch(service_info->mitm_query_h));
+            }
+
+            /* Parse response to see if we should mitm. */
+            bool should_mitm;
+            {
+                IpcParsedCommand r;
+                ipcParse(&r);
+                struct {
+                    u64 magic;
+                    u64 result;
+                    bool should_mitm;
+                } *resp = ((decltype(resp))r.Raw);
+
+                R_TRY(resp->result);
+                should_mitm = resp->should_mitm;
+            }
+
+            /* If we shouldn't mitm, give normal session. */
+            if (!should_mitm) {
+                return svcConnectToPort(out, service_info->port_h);
+            }
+
+            /* Create both handles. If the second fails, close the first to prevent leak. */
+            R_TRY(svcConnectToPort(&service_info->mitm_fwd_sess_h, service_info->port_h));
+            R_TRY_CLEANUP(svcConnectToPort(out, service_info->mitm_port_h), {
+                svcCloseHandle(service_info->mitm_fwd_sess_h);
+                service_info->mitm_fwd_sess_h = 0;
+            });
+
+            service_info->mitm_waiting_ack_pid = pid;
+            service_info->mitm_waiting_ack = true;
+
+            return ResultSuccess;
+        }
+
+        Result GetServiceHandleImpl(Handle *out, ServiceInfo *service_info, u64 pid) {
+            /* Clear handle output. */
+            *out = 0;
+
+            /* If not mitm'd or mitm service is requesting, get normal session. */
+            if (!IsValidProcessId(service_info->mitm_pid) || service_info->mitm_pid == pid) {
+                return svcConnectToPort(out, service_info->port_h);
+            }
+
+            /* We're mitm'd. */
+            if (R_FAILED(GetMitmServiceHandleImpl(out, service_info, pid))) {
+                /* If the Mitm service is dead, just give a normal session. */
+                return svcConnectToPort(out, service_info->port_h);
+            }
+
+            return ResultSuccess;
+        }
+
+        Result RegisterServiceImpl(Handle *out, u64 pid, ServiceName service, size_t max_sessions, bool is_light) {
+            /* Validate service name. */
+            R_TRY(ValidateServiceName(service));
+
+            /* Don't try to register something already registered. */
+            if (HasServiceInfo(service)) {
+                return ResultSmAlreadyRegistered;
+            }
+
+            /* Adjust session limit, if compile flags tell us to. */
+#ifdef SM_MINIMUM_SESSION_LIMIT
```

Comment:
One of the things that the Linux kernel does is avoids `#ifdef` in places like these in favor of `if(SM_MINIMUM_SESSION_LIMIT) {`, since the compiler can just remove the code if it's false, but then if someone breaks the code inside, you still get warnings and errors. Maybe we should do that too?